// @glimmer/validator/dist/commonjs/es2017/lib/tracking.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.EPOCH = undefined;
exports.track = track;
exports.consume = consume;
exports.trackedData = trackedData;

var _validators = require('./validators');

var _meta = require('./meta');

/**
 * Whenever a tracked computed property is entered, the current tracker is
 * saved off and a new tracker is replaced.
 *
 * Any tracked properties consumed are added to the current tracker.
 *
 * When a tracked computed property is exited, the tracker's tags are
 * combined and added to the parent tracker.
 *
 * The consequence is that each tracked computed property has a tag
 * that corresponds to the tracked properties consumed inside of
 * itself, including child tracked computed properties.
 */
let CURRENT_TRACKER = null;
/**
 * An object that that tracks @tracked properties that were consumed.
 */
class Tracker {
    constructor() {
        this.tags = new Set();
        this.last = null;
    }
    add(tag) {
        this.tags.add(tag);
        this.last = tag;
    }
    combine() {
        let { tags } = this;
        if (tags.size === 0) {
            return _validators.CONSTANT_TAG;
        } else if (tags.size === 1) {
            return this.last;
        } else {
            let tagsArr = [];
            tags.forEach(tag => tagsArr.push(tag));
            return (0, _validators.combine)(tagsArr);
        }
    }
}
function track(callback) {
    let parent = CURRENT_TRACKER;
    let current = new Tracker();
    CURRENT_TRACKER = current;
    try {
        callback();
    } finally {
        CURRENT_TRACKER = parent;
    }
    return current.combine();
}
function consume(tag) {
    if (CURRENT_TRACKER !== null) {
        CURRENT_TRACKER.add(tag);
    }
}
//////////
const EPOCH = exports.EPOCH = (0, _validators.createTag)();
function trackedData(key, initializer) {
    let values = new WeakMap();
    let hasInitializer = typeof initializer === 'function';
    function getter(self) {
        consume((0, _meta.tagFor)(self, key));
        let value;
        // If the field has never been initialized, we should initialize it
        if (hasInitializer && !values.has(self)) {
            value = initializer();
            values.set(self, value);
        } else {
            value = values.get(self);
        }
        return value;
    }
    function setter(self, value) {
        (0, _validators.dirty)(EPOCH);
        (0, _meta.dirtyTag)(self, key);
        values.set(self, value);
    }
    return { getter, setter };
}
