// @glimmer/validator/dist/amd/es5/glimmer-validator.js
define('@glimmer/validator', ['exports'], function (exports) { 'use strict';

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

    var symbol = typeof Symbol !== 'undefined' ? Symbol : function (key) {
        return '__' + key + Math.floor(Math.random() * Date.now()) + '__';
    };
    var CONSTANT = 0;
    var INITIAL = 1;
    var VOLATILE = 9007199254740991; // MAX_INT
    var $REVISION = INITIAL;
    function bump() {
        $REVISION++;
    }
    //////////
    var COMPUTE = symbol('TAG_COMPUTE');
    //////////
    /**
     * `value` receives a tag and returns an opaque Revision based on that tag. This
     * snapshot can then later be passed to `validate` with the same tag to
     * determine if the tag has changed at all since the time that `value` was
     * called.
     *
     * The current implementation returns the global revision count directly for
     * performance reasons. This is an implementation detail, and should not be
     * relied on directly by users of these APIs. Instead, Revisions should be
     * treated as if they are opaque/unknown, and should only be interacted with via
     * the `value`/`validate` API.
     *
     * @param tag
     */
    function value(_tag) {
        return $REVISION;
    }
    /**
     * `validate` receives a tag and a snapshot from a previous call to `value` with
     * the same tag, and determines if the tag is still valid compared to the
     * snapshot. If the tag's state has changed at all since then, `validate` will
     * return false, otherwise it will return true. This is used to determine if a
     * calculation related to the tags should be rerun.
     *
     * @param tag
     * @param snapshot
     */
    function validate(tag, snapshot) {
        return snapshot >= tag[COMPUTE]();
    }
    var TYPE = symbol('TAG_TYPE');
    var ALLOW_CYCLES = void 0;

    var MonomorphicTagImpl = function () {
        function MonomorphicTagImpl(type) {
            _classCallCheck(this, MonomorphicTagImpl);

            this.revision = INITIAL;
            this.lastChecked = INITIAL;
            this.lastValue = INITIAL;
            this.isUpdating = false;
            this.subtag = null;
            this.subtags = null;
            this[TYPE] = type;
        }

        MonomorphicTagImpl.prototype[COMPUTE] = function () {
            var lastChecked = this.lastChecked;

            if (lastChecked !== $REVISION) {
                this.isUpdating = true;
                this.lastChecked = $REVISION;
                try {
                    var subtags = this.subtags,
                        subtag = this.subtag,
                        revision = this.revision;

                    if (subtag !== null) {
                        revision = Math.max(revision, subtag[COMPUTE]());
                    }
                    if (subtags !== null) {
                        for (var i = 0; i < subtags.length; i++) {
                            var _value = subtags[i][COMPUTE]();
                            revision = Math.max(_value, revision);
                        }
                    }
                    this.lastValue = revision;
                } finally {
                    this.isUpdating = false;
                }
            }
            if (this.isUpdating === true) {
                this.lastChecked = ++$REVISION;
            }
            return this.lastValue;
        };

        MonomorphicTagImpl.update = function update(_tag, subtag) {
            // TODO: TS 3.7 should allow us to do this via assertion
            var tag = _tag;
            if (subtag === CONSTANT_TAG) {
                tag.subtag = null;
            } else {
                tag.subtag = subtag;
                // subtag could be another type of tag, e.g. CURRENT_TAG or VOLATILE_TAG.
                // If so, lastChecked/lastValue will be undefined, result in these being
                // NaN. This is fine, it will force the system to recompute.
                tag.lastChecked = Math.min(tag.lastChecked, subtag.lastChecked);
                tag.lastValue = Math.max(tag.lastValue, subtag.lastValue);
            }
        };

        MonomorphicTagImpl.dirty = function dirty(tag) {
            tag.revision = ++$REVISION;
        };

        return MonomorphicTagImpl;
    }();

    var dirty = MonomorphicTagImpl.dirty;
    var update = MonomorphicTagImpl.update;
    //////////
    function createTag() {
        return new MonomorphicTagImpl(0 /* Dirtyable */);
    }
    function createUpdatableTag() {
        return new MonomorphicTagImpl(1 /* Updatable */);
    }
    //////////
    var CONSTANT_TAG = new MonomorphicTagImpl(3 /* Constant */);
    function isConst(_ref) {
        var tag = _ref.tag;

        return tag === CONSTANT_TAG;
    }
    function isConstTag(tag) {
        return tag === CONSTANT_TAG;
    }
    //////////

    var VolatileTag = function () {
        function VolatileTag() {
            _classCallCheck(this, VolatileTag);
        }

        VolatileTag.prototype[COMPUTE] = function () {
            return VOLATILE;
        };

        return VolatileTag;
    }();

    var VOLATILE_TAG = new VolatileTag();
    //////////

    var CurrentTag = function () {
        function CurrentTag() {
            _classCallCheck(this, CurrentTag);
        }

        CurrentTag.prototype[COMPUTE] = function () {
            return $REVISION;
        };

        return CurrentTag;
    }();

    var CURRENT_TAG = new CurrentTag();
    //////////
    function combine(tags) {
        var optimized = [];
        for (var i = 0, l = tags.length; i < l; i++) {
            var tag = tags[i];
            if (tag === CONSTANT_TAG) continue;
            optimized.push(tag);
        }
        return createCombinatorTag(optimized);
    }
    function createCombinatorTag(tags) {
        switch (tags.length) {
            case 0:
                return CONSTANT_TAG;
            case 1:
                return tags[0];
            default:
                var tag = new MonomorphicTagImpl(2 /* Combinator */);
                tag.subtags = tags;
                return tag;
        }
    }

    var TRACKED_TAGS = new WeakMap();
    function isObject(u) {
        return typeof u === 'object' && u !== null;
    }
    function dirtyTag(obj, key) {
        if (isObject(obj)) {
            var tag = tagFor(obj, key);
            if (tag === undefined) {
                updateTag(obj, key, createUpdatableTag());
            } else if (isConstTag(tag)) {
                throw new Error('BUG: Can\'t update a constant tag');
            } else {
                dirty(tag);
            }
        } else {
            throw new Error('BUG: Can\'t update a tag for a primitive');
        }
    }
    function tagFor(obj, key) {
        if (isObject(obj)) {
            var tags = TRACKED_TAGS.get(obj);
            if (tags === undefined) {
                tags = new Map();
                TRACKED_TAGS.set(obj, tags);
            } else if (tags.has(key)) {
                return tags.get(key);
            }
            var tag = createUpdatableTag();
            tags.set(key, tag);
            return tag;
        } else {
            return CONSTANT_TAG;
        }
    }
    function updateTag(obj, key, newTag) {
        if (isObject(obj)) {
            var tag = tagFor(obj, key);
            if (isConstTag(tag)) {
                throw new Error('BUG: Can\'t update a constant tag');
            } else {
                update(tag, newTag);
            }
            return tag;
        } else {
            throw new Error('BUG: Can\'t update a tag for a primitive');
        }
    }

    function _classCallCheck$1(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
    /**
     * Whenever a tracked computed property is entered, the current tracker is
     * saved off and a new tracker is replaced.
     *
     * Any tracked properties consumed are added to the current tracker.
     *
     * When a tracked computed property is exited, the tracker's tags are
     * combined and added to the parent tracker.
     *
     * The consequence is that each tracked computed property has a tag
     * that corresponds to the tracked properties consumed inside of
     * itself, including child tracked computed properties.
     */
    var CURRENT_TRACKER = null;
    /**
     * An object that that tracks @tracked properties that were consumed.
     */

    var Tracker = function () {
        function Tracker() {
            _classCallCheck$1(this, Tracker);

            this.tags = new Set();
            this.last = null;
        }

        Tracker.prototype.add = function add(tag) {
            this.tags.add(tag);
            this.last = tag;
        };

        Tracker.prototype.combine = function combine$$1() {
            var tags = this.tags;

            if (tags.size === 0) {
                return CONSTANT_TAG;
            } else if (tags.size === 1) {
                return this.last;
            } else {
                var tagsArr = [];
                tags.forEach(function (tag) {
                    return tagsArr.push(tag);
                });
                return combine(tagsArr);
            }
        };

        return Tracker;
    }();

    function track(callback) {
        var parent = CURRENT_TRACKER;
        var current = new Tracker();
        CURRENT_TRACKER = current;
        try {
            callback();
        } finally {
            CURRENT_TRACKER = parent;
        }
        return current.combine();
    }
    function consume(tag) {
        if (CURRENT_TRACKER !== null) {
            CURRENT_TRACKER.add(tag);
        }
    }
    //////////
    var EPOCH = createTag();
    function trackedData(key, initializer) {
        var values = new WeakMap();
        var hasInitializer = typeof initializer === 'function';
        function getter(self) {
            consume(tagFor(self, key));
            var value$$1 = void 0;
            // If the field has never been initialized, we should initialize it
            if (hasInitializer && !values.has(self)) {
                value$$1 = initializer();
                values.set(self, value$$1);
            } else {
                value$$1 = values.get(self);
            }
            return value$$1;
        }
        function setter(self, value$$1) {
            dirty(EPOCH);
            dirtyTag(self, key);
            values.set(self, value$$1);
        }
        return { getter: getter, setter: setter };
    }

    exports.ALLOW_CYCLES = ALLOW_CYCLES;
    exports.bump = bump;
    exports.combine = combine;
    exports.COMPUTE = COMPUTE;
    exports.CONSTANT_TAG = CONSTANT_TAG;
    exports.CONSTANT = CONSTANT;
    exports.createCombinatorTag = createCombinatorTag;
    exports.createTag = createTag;
    exports.createUpdatableTag = createUpdatableTag;
    exports.CURRENT_TAG = CURRENT_TAG;
    exports.dirty = dirty;
    exports.INITIAL = INITIAL;
    exports.isConst = isConst;
    exports.isConstTag = isConstTag;
    exports.update = update;
    exports.validate = validate;
    exports.value = value;
    exports.VOLATILE_TAG = VOLATILE_TAG;
    exports.VOLATILE = VOLATILE;
    exports.dirtyTag = dirtyTag;
    exports.tagFor = tagFor;
    exports.updateTag = updateTag;
    exports.track = track;
    exports.consume = consume;
    exports.EPOCH = EPOCH;
    exports.trackedData = trackedData;

    Object.defineProperty(exports, '__esModule', { value: true });

});
