// @glimmer/validator/dist/modules/es5/lib/tracking.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

import { combine as _combine, CONSTANT_TAG } from './validators';
import { createTag, dirty } from './validators';
import { tagFor, dirtyTag } from './meta';
/**
 * Whenever a tracked computed property is entered, the current tracker is
 * saved off and a new tracker is replaced.
 *
 * Any tracked properties consumed are added to the current tracker.
 *
 * When a tracked computed property is exited, the tracker's tags are
 * combined and added to the parent tracker.
 *
 * The consequence is that each tracked computed property has a tag
 * that corresponds to the tracked properties consumed inside of
 * itself, including child tracked computed properties.
 */
var CURRENT_TRACKER = null;
/**
 * An object that that tracks @tracked properties that were consumed.
 */

var Tracker = function () {
    function Tracker() {
        _classCallCheck(this, Tracker);

        this.tags = new Set();
        this.last = null;
    }

    Tracker.prototype.add = function add(tag) {
        this.tags.add(tag);
        this.last = tag;
    };

    Tracker.prototype.combine = function combine() {
        var tags = this.tags;

        if (tags.size === 0) {
            return CONSTANT_TAG;
        } else if (tags.size === 1) {
            return this.last;
        } else {
            var tagsArr = [];
            tags.forEach(function (tag) {
                return tagsArr.push(tag);
            });
            return _combine(tagsArr);
        }
    };

    return Tracker;
}();

export function track(callback) {
    var parent = CURRENT_TRACKER;
    var current = new Tracker();
    CURRENT_TRACKER = current;
    try {
        callback();
    } finally {
        CURRENT_TRACKER = parent;
    }
    return current.combine();
}
export function consume(tag) {
    if (CURRENT_TRACKER !== null) {
        CURRENT_TRACKER.add(tag);
    }
}
//////////
export var EPOCH = createTag();
export function trackedData(key, initializer) {
    var values = new WeakMap();
    var hasInitializer = typeof initializer === 'function';
    function getter(self) {
        consume(tagFor(self, key));
        var value = void 0;
        // If the field has never been initialized, we should initialize it
        if (hasInitializer && !values.has(self)) {
            value = initializer();
            values.set(self, value);
        } else {
            value = values.get(self);
        }
        return value;
    }
    function setter(self, value) {
        dirty(EPOCH);
        dirtyTag(self, key);
        values.set(self, value);
    }
    return { getter: getter, setter: setter };
}
