// @glimmer/reference/dist/commonjs/es2017/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "REFERENCE", {
  enumerable: true,
  get: function () {
    return _reference.REFERENCE;
  }
});
Object.defineProperty(exports, "createPrimitiveRef", {
  enumerable: true,
  get: function () {
    return _reference.createPrimitiveRef;
  }
});
Object.defineProperty(exports, "createConstRef", {
  enumerable: true,
  get: function () {
    return _reference.createConstRef;
  }
});
Object.defineProperty(exports, "createUnboundRef", {
  enumerable: true,
  get: function () {
    return _reference.createUnboundRef;
  }
});
Object.defineProperty(exports, "createComputeRef", {
  enumerable: true,
  get: function () {
    return _reference.createComputeRef;
  }
});
Object.defineProperty(exports, "createDebugAliasRef", {
  enumerable: true,
  get: function () {
    return _reference.createDebugAliasRef;
  }
});
Object.defineProperty(exports, "createReadOnlyRef", {
  enumerable: true,
  get: function () {
    return _reference.createReadOnlyRef;
  }
});
Object.defineProperty(exports, "createInvokableRef", {
  enumerable: true,
  get: function () {
    return _reference.createInvokableRef;
  }
});
Object.defineProperty(exports, "isInvokableRef", {
  enumerable: true,
  get: function () {
    return _reference.isInvokableRef;
  }
});
Object.defineProperty(exports, "isConstRef", {
  enumerable: true,
  get: function () {
    return _reference.isConstRef;
  }
});
Object.defineProperty(exports, "isUpdatableRef", {
  enumerable: true,
  get: function () {
    return _reference.isUpdatableRef;
  }
});
Object.defineProperty(exports, "valueForRef", {
  enumerable: true,
  get: function () {
    return _reference.valueForRef;
  }
});
Object.defineProperty(exports, "updateRef", {
  enumerable: true,
  get: function () {
    return _reference.updateRef;
  }
});
Object.defineProperty(exports, "childRefFor", {
  enumerable: true,
  get: function () {
    return _reference.childRefFor;
  }
});
Object.defineProperty(exports, "childRefFromParts", {
  enumerable: true,
  get: function () {
    return _reference.childRefFromParts;
  }
});
Object.defineProperty(exports, "UNDEFINED_REFERENCE", {
  enumerable: true,
  get: function () {
    return _reference.UNDEFINED_REFERENCE;
  }
});
Object.defineProperty(exports, "NULL_REFERENCE", {
  enumerable: true,
  get: function () {
    return _reference.NULL_REFERENCE;
  }
});
Object.defineProperty(exports, "TRUE_REFERENCE", {
  enumerable: true,
  get: function () {
    return _reference.TRUE_REFERENCE;
  }
});
Object.defineProperty(exports, "FALSE_REFERENCE", {
  enumerable: true,
  get: function () {
    return _reference.FALSE_REFERENCE;
  }
});
Object.defineProperty(exports, "createIteratorRef", {
  enumerable: true,
  get: function () {
    return _iterable.createIteratorRef;
  }
});
Object.defineProperty(exports, "createIteratorItemRef", {
  enumerable: true,
  get: function () {
    return _iterable.createIteratorItemRef;
  }
});

var _reference = require("./lib/reference");

var _iterable = require("./lib/iterable");
