// @glimmer/util/dist/commonjs/es2017/lib/platform-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keys = keys;
exports.unwrap = unwrap;
exports.expect = expect;
exports.unreachable = unreachable;
exports.exhausted = exhausted;
exports.symbol = exports.tuple = exports.HAS_NATIVE_SYMBOL = void 0;

const HAS_NATIVE_SYMBOL = function () {
  if (typeof Symbol !== 'function') {
    return false;
  } // eslint-disable-next-line symbol-description


  return typeof Symbol() === 'symbol';
}();

exports.HAS_NATIVE_SYMBOL = HAS_NATIVE_SYMBOL;

function keys(obj) {
  return Object.keys(obj);
}

function unwrap(val) {
  if (val === null || val === undefined) throw new Error(`Expected value to be present`);
  return val;
}

function expect(val, message) {
  if (val === null || val === undefined) throw new Error(message);
  return val;
}

function unreachable(message = 'unreachable') {
  return new Error(message);
}

function exhausted(value) {
  throw new Error(`Exhausted ${value}`);
}

const tuple = (...args) => args;

exports.tuple = tuple;
const symbol = HAS_NATIVE_SYMBOL ? Symbol : key => `__${key}${Math.floor(Math.random() * Date.now())}__`;
exports.symbol = symbol;
