// @glimmer/validator/dist/commonjs/es2017/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _validators = require('./lib/validators');

Object.defineProperty(exports, 'ALLOW_CYCLES', {
  enumerable: true,
  get: function () {
    return _validators.ALLOW_CYCLES;
  }
});
Object.defineProperty(exports, 'bump', {
  enumerable: true,
  get: function () {
    return _validators.bump;
  }
});
Object.defineProperty(exports, 'combine', {
  enumerable: true,
  get: function () {
    return _validators.combine;
  }
});
Object.defineProperty(exports, 'COMPUTE', {
  enumerable: true,
  get: function () {
    return _validators.COMPUTE;
  }
});
Object.defineProperty(exports, 'CONSTANT_TAG', {
  enumerable: true,
  get: function () {
    return _validators.CONSTANT_TAG;
  }
});
Object.defineProperty(exports, 'CONSTANT', {
  enumerable: true,
  get: function () {
    return _validators.CONSTANT;
  }
});
Object.defineProperty(exports, 'createCombinatorTag', {
  enumerable: true,
  get: function () {
    return _validators.createCombinatorTag;
  }
});
Object.defineProperty(exports, 'createTag', {
  enumerable: true,
  get: function () {
    return _validators.createTag;
  }
});
Object.defineProperty(exports, 'createUpdatableTag', {
  enumerable: true,
  get: function () {
    return _validators.createUpdatableTag;
  }
});
Object.defineProperty(exports, 'CURRENT_TAG', {
  enumerable: true,
  get: function () {
    return _validators.CURRENT_TAG;
  }
});
Object.defineProperty(exports, 'dirty', {
  enumerable: true,
  get: function () {
    return _validators.dirty;
  }
});
Object.defineProperty(exports, 'INITIAL', {
  enumerable: true,
  get: function () {
    return _validators.INITIAL;
  }
});
Object.defineProperty(exports, 'isConst', {
  enumerable: true,
  get: function () {
    return _validators.isConst;
  }
});
Object.defineProperty(exports, 'isConstTag', {
  enumerable: true,
  get: function () {
    return _validators.isConstTag;
  }
});
Object.defineProperty(exports, 'update', {
  enumerable: true,
  get: function () {
    return _validators.update;
  }
});
Object.defineProperty(exports, 'validate', {
  enumerable: true,
  get: function () {
    return _validators.validate;
  }
});
Object.defineProperty(exports, 'value', {
  enumerable: true,
  get: function () {
    return _validators.value;
  }
});
Object.defineProperty(exports, 'VOLATILE_TAG', {
  enumerable: true,
  get: function () {
    return _validators.VOLATILE_TAG;
  }
});
Object.defineProperty(exports, 'VOLATILE', {
  enumerable: true,
  get: function () {
    return _validators.VOLATILE;
  }
});

var _meta = require('./lib/meta');

Object.defineProperty(exports, 'dirtyTag', {
  enumerable: true,
  get: function () {
    return _meta.dirtyTag;
  }
});
Object.defineProperty(exports, 'tagFor', {
  enumerable: true,
  get: function () {
    return _meta.tagFor;
  }
});
Object.defineProperty(exports, 'updateTag', {
  enumerable: true,
  get: function () {
    return _meta.updateTag;
  }
});

var _tracking = require('./lib/tracking');

Object.defineProperty(exports, 'track', {
  enumerable: true,
  get: function () {
    return _tracking.track;
  }
});
Object.defineProperty(exports, 'consume', {
  enumerable: true,
  get: function () {
    return _tracking.consume;
  }
});
Object.defineProperty(exports, 'EPOCH', {
  enumerable: true,
  get: function () {
    return _tracking.EPOCH;
  }
});
Object.defineProperty(exports, 'trackedData', {
  enumerable: true,
  get: function () {
    return _tracking.trackedData;
  }
});
